### PR TITLE
Array push optimization

### PR DESCRIPTION
**Pull request by marchant**

### CODE REVIEW

File: listen/array-changes.js [hunk 1]
```diff
@@ -162,18 +187,12 @@ var observableArrayProperties = {
...
+                this._dispatchBeforeOwnPropertyChange(start, plus.length);
+            } else if ((hasOwnPropertyChangeDescriptor = PropertyChanges.hasOwnPropertyChangeDescriptor(this))) {
                 // all subsequent values changed or shifted.
                 // avoid (longest - start) long walks if there are no
                 // registered descriptors.
-                for (var i = start; i < longest; i++) {
-                    PropertyChanges.dispatchBeforeOwnPropertyChange(this, i, this[i]);
-                    this.dispatchBeforeMapChange(i, this[i]);
-                }
+				this._dispatchBeforeOwnPropertyChange(start, longest-start);
```
Comment (kriskowal) [inline listen/array-changes.js]: Expand indents to four spaces.

File: listen/array-changes.js [hunk 2]
```diff
@@ -104,7 +110,7 @@ var observableArrayProperties = {
             array_sort.apply(this, arguments);
 
             // dispatch after change events
-            for (var i = 0; i < this.length; i++) {
+            for (i = 0, countI = this.length; i < countI; i++) {
```
Comment (kriskowal) [inline listen/array-changes.js]: I’ve used the pattern elsewhere: `for (var index = 0, length = this.length; index < length; index++)` in keeping with using full and meaningful names.

File: listen/array-changes.js [hunk 3]
```diff
@@ -254,12 +267,33 @@ var observableArrayProperties = {
...
+			array_push.apply(this,arguments);
+
+	        if (addedCount > 0) {
+				if (hasOwnPropertyChangeDescriptor) {
+					this._dispatchOwnPropertyChange(start,addedCount);
+				}
+		        this.dispatchRangeChange(argArray,Array.empty, start);
+				this.dispatchOwnPropertyChange("length", this.length);
+	        }
+
```
Comment (kriskowal) [inline listen/array-changes.js]: This is a reasonable case to unroll. In the v2 branch, I’m toying with changing the signature of swap to be the better kernel for all mutations, as documented here http://github.com/kriskowal/pop-swap

**Comment by kriskowal** [comment]
Apart from the indentation, looks good to me. All other comments should not block.

**Comment by marchant** [comment]
Kris, should be good to go, thanks for looking it over

**Comment by kriskowal** [comment]
Looks fine. Would be good to follow-up with a pass over with JSHint.

**Pull request merged**